File: lib/karabinex/keymap.ex
defmodule Karabinex.Keymap do
  alias Karabinex.{Command, Chord}

  defstruct [:chord, :hook, children: []]

  @type binding :: atom() | String.t()
  @type spec :: %{binding() => Command.spec() | spec()}

  @type t :: %__MODULE__{
          chord: Chord.t(),
          children: [t() | Command.t()]
        }

  @spec new(Chord.t(), [t() | Command.t()]) :: t()
  def new(chord, children) do
    %__MODULE__{
      chord: chord,
      children: children
    }
  end

  @spec add_hook(t(), Command.t()) :: t()
  def add_hook(%__MODULE__{} = keymap, %Command{} = hook) do
    %{keymap | hook: hook}
  end
end

File: src/keymap.ts
import type { Chord } from "./chord";
import type { Command } from "./command";

export class Keymap {
  readonly chord: Chord;
  readonly children: ReadonlyArray<Keymap | Command>;
  hook?: Command;

  constructor(chord: Chord, children: ReadonlyArray<Keymap | Command>) {
    this.chord = chord;
    this.children = children;
  }

  addHook(hook: Command): void {
    this.hook = hook;
  }
}
